### PR TITLE
feat: `Provider::subscribe_logs`

### PR DESCRIPTION
**Pull request by klkvr**
## Motivation

Adds `Provider::subscribe_logs` and `Event::subscribe` methods

## PR Checklist

- [ ] Added Tests
- [ ] Added Documentation
- [ ] Breaking changes

### CODE REVIEW

**Review by onbjerg — CHANGES_REQUESTED** [review]
lgtm, I would want `Event::subscribe` too

**Review by onbjerg — APPROVED** [review]
lgtm

**Review by mattsse — APPROVED** [review]
lgtm

pending @DaniPopes

**Review by prestwich — COMMENTED** [review]
@DaniPopes can we somehow unify the usage of `Subscription<T>` and `PollChannel<T>` to have a single front-end that works for both polling and pubsub?

File: crates/contract/src/event.rs [hunk 1]
```diff
@@ -154,7 +220,7 @@ mod tests {
     #[tokio::test]
     async fn event_filters() {
         init_tracing();
-        let (provider, _anvil) = spawn_anvil();
+        let (provider, anvil) = spawn_anvil();
```
Comment (DaniPopes) [inline crates/contract/src/event.rs:223]: Keep underscore for `not(feature = "pubsub")`

```suggestion
        let (provider, _anvil) = spawn_anvil();
```

**Review by DaniPopes — APPROVED** [review]

**Comment by DaniPopes** [comment]
@prestwich I think so, but I feel that would complicate things quite a bit for not much benefit?

**Comment by prestwich** [comment]
it means that users would follow the same API regardless of provider and get automatically upgraded to pubsub subscriptions when they change endpoints (unless they specifically want to downgrade to polling for some reason)

**Comment by onbjerg** [comment]
i agree w prestwich here, this would most likely also give us transport stacking back?

File: crates/contract/src/event.rs [hunk 2]
```diff
@@ -131,10 +140,67 @@ fn decode_log<E: SolEvent>(log: &Log) -> alloy_sol_types::Result<E> {
...
+                .field("event_type", &format_args!("{}", std::any::type_name::<E>()))
+                .finish()
+        }
+    }
+
+    impl<E: SolEvent> EventSubscription<E> {
+        /// Creates a new event subscription with the provided subscription.
+        #[allow(clippy::missing_const_for_fn)]
+        #[inline]
+        pub fn new(sub: Subscription<Log>) -> Self {
```
Comment (prestwich) [inline crates/contract/src/event.rs:186]: `From` impl preferred here

File: crates/contract/src/event.rs [hunk 3]
```diff
@@ -131,10 +140,67 @@ fn decode_log<E: SolEvent>(log: &Log) -> alloy_sol_types::Result<E> {
...
+            Self { sub, _phantom: PhantomData }
+        }
+
+        /// Converts the subscription into a stream.
```
Comment (prestwich) [inline crates/contract/src/event.rs:190]: This results in an unnameable type. Is there any situation in which a user would want to name the resulting stream?

**Review by prestwich — CHANGES_REQUESTED** [review]

**Comment by prestwich** [comment]
Regardless, that's not in scope for this PR. Opened #346 to capture.

File: crates/contract/src/event.rs [hunk 4]
```diff
@@ -131,10 +140,67 @@ fn decode_log<E: SolEvent>(log: &Log) -> alloy_sol_types::Result<E> {
...
+    /// An event subscription.
+    ///
+    /// Underlying subscription is available through the [`sub`](Self::sub) field.
+    pub struct EventSubscription<E> {
+        /// The inner poller.
+        pub sub: Subscription<Log>,
+        _phantom: PhantomData<E>,
+    }
+
+    impl<E> std::ops::Deref for EventSubscription<E> {
```
[the remark this answers is not in the document]
Comment (klkvr) [reply crates/contract/src/event.rs:157]: I think we can, @DaniPopes wdyt?

File: crates/contract/src/event.rs (hunk 2 above)
Comment (klkvr) [reply crates/contract/src/event.rs]: added for both `EventPoller` and `EventSubscription` in d37f0503918e7a7d9e953e664c3b4293f25a1bad

File: crates/contract/src/event.rs (hunk 3 above)
Comment (mattsse) [reply crates/contract/src/event.rs:188]: I guess we want a similar thing as `SubscriptionStream` here with the type we decode on poll_next?

Comment (prestwich) [reply crates/contract/src/event.rs:188]: ideally we have these all work the same way. can convert this to issue?

File: crates/contract/Cargo.toml [hunk 5]
```diff
@@ -26,10 +26,15 @@ futures-util.workspace = true
...
-alloy-rpc-client.workspace = true
+alloy-rpc-client = { workspace = true, features = ["pubsub", "ws"] }
 alloy-transport-http.workspace = true
 alloy-node-bindings.workspace = true
 test-utils.workspace = true
 reqwest.workspace = true
 tokio = { workspace = true, features = ["macros", "rt-multi-thread"] }
+
+[features]
+pubsub = ["alloy-provider/pubsub", "dep:alloy-pubsub"]
```
Comment (prestwich) [inline crates/contract/Cargo.toml:40]: nit: should have a newline

File: crates/contract/src/event.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (DaniPopes) [reply crates/contract/src/event.rs:156]: no real reason other than allowing `x.set_thing` rather than `x.inner.set_thing`

**Comment by prestwich** [comment]
re-running CI before merge

**Pull request merged**